modular_skyrat/modules/modular_implants/code/nifs.dm

```diff
@@ -0,0 +1,511 @@
+	///This is the sound that you would hear if you enable if you activate or enable something.
+	var/click_sound = 'modular_skyrat/modules/modular_implants/sounds/default_click.ogg'
+	///What icon does the NIF display in chat when sending out alerts? Icon states are stored in 'modular_skyrat/modules/modular_implants/icons/chat.dmi'
+	var/chat_icon = "standard"
+
+/obj/item/organ/internal/cyberimp/brain/nif/Initialize(mapload)
+	. = ..()
+
+	durability = max_durability
+	loaded_nifsofts = preinstalled_nifsofts
+
+	for(var/datum/nifsoft/preinstalled_nifsoft as anything in preinstalled_nifsofts)
+		install_nifsoft(preinstalled_nifsoft)
+
+	power_level = max_power
+
+/obj/item/organ/internal/cyberimp/brain/nif/Insert(mob/living/carbon/human/insertee)
+	. = ..()
+
+	if(linked_mob && stored_ckey != insertee.ckey && theft_protection)
+		insertee.audible_message(span_warning("The [src] lets out a negative buzz before forcefully removing itself from [insertee]'s brain."))
+		playsound(insertee, 'sound/machines/buzz-sigh.ogg', 30, TRUE)
+		src.Remove(insertee)
+		src.forceMove(get_turf(insertee))
+
+		return FALSE
+
+	linked_mob = insertee
+	stored_ckey = linked_mob.ckey
+
+	loc = insertee // This needs to be done, otherwise TGUI will not pull up.
+	insertee.installed_nif = src
+	START_PROCESSING(SSobj, src)
+
+	if(!is_calibrated)
+		send_message("Calibration Process Starting!")
+		calibrating = TRUE
+
+/obj/item/organ/internal/cyberimp/brain/nif/Remove(mob/living/carbon/organ_owner, special = FALSE)
+	. = ..()
+
+	organ_owner.log_message("[src] was removed from [organ_owner]]",LOG_GAME)
+	STOP_PROCESSING(SSobj, src)
+
+/obj/item/organ/internal/cyberimp/brain/nif/process(delta_time)
+	. = ..()
+
+	if(!linked_mob || broken)
+		return FALSE
+
+	if(calibrating)
+		preform_calibration()
+
+	if(IS_IN_STASIS(linked_mob))
+		return
+
+	if((linked_mob.nutrition < minimum_nutrition) && (nutrition_drain)) //Turns nutrition drain off if nutrition is lower than minimum
+		toggle_nutrition_drain(TRUE)
+
+	if(blood_drain && (!blood_check())) //Disables blood draining if the mob fails the blood check
+		toggle_blood_drain(TRUE)
+
+	if(blood_drain)
+		linked_mob.blood_volume -= blood_drain_rate
+
+	if(power_usage > power_level)
+		for(var/datum/nifsoft/nifsoft as anything in loaded_nifsofts)
+			if(!nifsoft.active)
+				continue
+
+			nifsoft.activate()
+
+	power_level += -power_usage
+	if(power_level > max_power)
+		power_level = max_power //No Overcharging
+
+/obj/item/organ/internal/cyberimp/brain/nif/Destroy()
+	linked_mob.nif_examine_text = null
+	linked_mob = null
+
+	QDEL_LIST(loaded_nifsofts)
+	return ..()
+
+/// Subtracts from the power level of the NIF once, this is good for anything that is single use.
+/obj/item/organ/internal/cyberimp/brain/nif/proc/use_power(power_to_use)
+	if(power_level < power_to_use)
+		return FALSE
+
+	power_level -= power_to_use
+	return TRUE
+
+///Toggles nutrition drain as a power source on NIFs on/off
```

Comment:
Should document what `bypass` is used for, here, as it's not clear that there's an implicit check in there.